### DISCUSSION TITLE
7.1. GraphQL schema should be "hairy" - how to avoid circular references?

### DISCUSSION THREAD

**Discussion by nareshbhatia** [graphql-rules/graphql-rules#124] [Q&A] [answered]
Based on this rule, I modeled a type called `User` which has a field called `friends`, which is an array of users:

```graphql
type User {
  id: ID!
  fullName: String!
  avatarUrl: String!
  homeTown: String!
  friends: [User!]!
}
```

However when I tried to mock a response using MSW, I got this error:

```
[MSW] Uncaught exception in the request handler for "POST http://localhost:3000/":

TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    |     property 'friends' -> object with constructor 'Array'
    |     index 0 -> object with constructor 'Object'
    |     property 'friends' -> object with constructor 'Array'
    --- index 0 closes the circle
```

How to avoid circular references while still making the schema hairy? I can only think about typing friends to be something like `UserSummary` which does not contain `friends`, but then one can't look up friends of friends!!

**Comment by RIP21** [comment, marked as answer]
This is rather complicated TBH. But you should build static manual responses using plain js objects to make sure they have finite depth. And better do that statically without referring to classes or object references that, I assume MSW serialiser tries to follow references all the way to the end that never happens as it's circular and non-serialisable.

A real query will have a finite depth, and so is the response.
